### ISSUE TITLE
Warn on accidental copying of arrays

### ISSUE THREAD

**Issue by aochagavia** [rust-lang/rust-clippy#1474]
The code below compiles correctly, but makes no sense at all. I think this violates the principle of least surprise (we had a nice discussion about it on #rust). See if you can spot the error ;)

```rust
extern crate scoped_threadpool;
fn main() {
    let mut array = [0; 3];
    let mut pool = scoped_threadpool::Pool::new(2);
    pool.scoped(|scope| {
        scope.execute(move||{
            array[0] = 1;
        });
        scope.execute(move||{
            array[0] = 2;
        });
    });
    // Prints 0, since the array is copied into each thread
    println!("{:?}",array[0]);
}
```
Would it be possible to add a lint for this? Just warning that a captured array of `Clone` elements will be copied and not moved.

Credits go to @arianvp for coming up with the code (and needing 5 days to discover why it was wrong)

**Comment by mcarton** [comment]
> we had a nice discussion about it on #rust

Do you have a link? Time of discussion? :smile:

**Comment by aochagavia** [comment]
According to the [botbot logs](https://botbot.me/mozilla/rust/2017-01-26/?msg=79964938&page=7) it was on 9:10 am today (January 26th 2017)

**Comment by camsteffen** [comment]
Just looking at the closures independently, you could lint that the (copy of the) array is modified and then dropped without ever observing the array after it was modified.

Otherwise, I don't think it is realistic to lint "accidental copy" because there is no reliable way of knowing if the copying is "accidental".

**Comment by camsteffen** [comment]
> Just looking at the closures independently, you could lint that the (copy of the) array is modified and then dropped without ever observing the array after it was modified.

This is rust-lang/rust#65467
